url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/Networking/Server.java

```diff
@@ -0,0 +1,502 @@
+import DeviceGraphics.DeviceGraphics;
+import DeviceGraphics.FeederGraphics;
+import DeviceGraphics.GantryGraphics;
+import DeviceGraphics.InspectionStandGraphics;
+import DeviceGraphics.KitRobotGraphics;
+import DeviceGraphics.LaneGraphics;
+import DeviceGraphics.NestGraphics;
+import DeviceGraphics.PartsRobotGraphics;
+import DeviceGraphics.StandGraphics;
+import Utils.ConsoleWriter;
+import Utils.Constants;
+import agent.Agent;
+import agent.CameraAgent;
+import agent.ConveyorAgent;
+import agent.FCSAgent;
+import agent.FeederAgent;
+import agent.GantryAgent;
+import agent.KitRobotAgent;
+import agent.LaneAgent;
+import agent.NestAgent;
+import agent.PartsRobotAgent;
+import agent.StandAgent;
+import factory.FCS;
+
+/**
+ * The Server is the "middleman" between Agents and the GUI clients. This is
+ * where constructors of the different Agents will be called, as well as
+ * establishing connections with the GUI clients.
+ * @author Peter Zhang
+ */
+public class Server {
+	private ServerSocket ss;
+	private Socket s;
+
+	// GUI client reader/writers
+	private ClientReader factProdMngrReader;
+	private AbstractWriter factProdMngrWriter = new DummyWriter();
+
+	private ClientReader partsMngrReader;
+	private AbstractWriter partsMngrWriter = new DummyWriter();
+
+	private ClientReader kitMngrReader;
+	private AbstractWriter kitMngrWriter = new DummyWriter();
+
+	// Graphics only client reader/writers
+	private ClientReader gantryRobotMngrReader;
+	private AbstractWriter gantryRobotMngrWriter = new DummyWriter();
+
+	private ClientReader kitAssemblyMngrReader;
+	private AbstractWriter kitAssemblyMngrWriter = new DummyWriter();
+
+	private ClientReader laneMngrReader;
+	private AbstractWriter laneMngrWriter = new DummyWriter();
+
+	// V0 Config (testing only - remove later)
+	private ClientReader kitRobotMngrReader;
+	private AbstractWriter kitRobotMngrWriter = new DummyWriter();
+
+	private ClientReader partsRobotMngrReader;
+	private AbstractWriter partsRobotMngrWriter = new DummyWriter();
+
+	// See how many clients have connected
+	private int numClients = 0;
+
+	public volatile LinkedHashMap<String, DeviceGraphics> devices = new LinkedHashMap<String, DeviceGraphics>();
+	public volatile LinkedHashMap<String, Agent> agents = new LinkedHashMap<String, Agent>();
+
+	FCS fcs;
+	FCSAgent fcsAgent;
+
+	private final ConsoleWriter console;
+
+	public Server() {
+		// connecting to cloud debug panel
+		console = new ConsoleWriter("Server");
+		console.startConsole();
+		console.startThread();
+
+		fcsAgent = new FCSAgent(Constants.FCS_TARGET);
+		fcs = new FCS(this, fcsAgent);
+
+		initAgents();
+		initDevices();
+		connectAgentsWithEachOther();
+		connectAgentsWithDevices();
+
+		// add the save file method (at bottom) to the server
+		addShutdownHook();
+
+		initStreams();
+
+		// will never run anything after init Streams
+	}
+
+	private void initStreams() {
+		try {
+			ss = new ServerSocket(Constants.SERVER_PORT);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *ss*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).